### PR TITLE
Muon - Widget; Load File Widget

### PR DESCRIPTION
**Pull request by TWJubb**
Refs #23642

Requires #23670 to be merged.

PR #23371 was moved here on request, so please see that PR for a more complete discussion. 

This PR creates the load file widget for the new Muon Analysis interface.

-------------------

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/load_file_widget/model.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import os
+
+from mantid.kernel import ConfigService
+from mantid.api import WorkspaceGroup
+
+from Muon.GUI.Common.muon_load_data import MuonLoadData
+import Muon.GUI.Common.utilities.load_utils as load_utils
+
+
+def is_workspace_group(workspace):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/model.py]: this is also in the previous PR (identical). So I think we should make some shared functionality.

File: scripts/Muon/GUI/Common/load_file_widget/model.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    return isinstance(workspace, WorkspaceGroup)
+
+
+def exception_message_for_failed_files(failed_file_list):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/model.py]: same

File: scripts/Muon/GUI/Common/load_file_widget/model.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    return "Could not load the following files : \n - " + "\n - ".join(failed_file_list)
+
+
+def get_run_from_multi_period_data(workspace_list):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/model.py]: shared

File: scripts/Muon/GUI/Common/load_file_widget/model.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        return self.loaded_runs
+
+    def add_directories_to_config_service(self, file_list):
+        """
+        Parses file_list into the unique directories containing the files, and adds these
+        to the global config service. These directories will then be automatically searched in
+        all subsequent Load calls.
+        """
+        dirs = [os.path.dirname(filename) for filename in file_list]
+        dirs = [filename if os.path.isdir(filename) else "" for filename in dirs]
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/model.py]: its not a filename at this point, shall we give it a more representative name?

File: scripts/Muon/GUI/Common/load_file_widget/presenter.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        else:
+            self.handle_load_no_threading(filenames)
+
+    def handle_load_no_threading(self, filenames):
+        self._view.notify_loading_started()
+        self.disable_loading()
+        self._model.loadData(filenames)
+        try:
+            self._model.execute()
+        except ValueError as e:
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/presenter.py]: error instead of e

File: scripts/Muon/GUI/Common/load_file_widget/view.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.browse_button = QtGui.QPushButton(self)
+
+        size_policy = QtGui.QSizePolicy(QtGui.QSizePolicy.Fixed, QtGui.QSizePolicy.Fixed)
+        size_policy.setHorizontalStretch(0)
+        size_policy.setVerticalStretch(0)
+        size_policy.setHeightForWidth(self.browse_button.sizePolicy().hasHeightForWidth())
+
+        self.browse_button.setSizePolicy(size_policy)
+        # self.browse_button.setMinimumSize(QtCore.QSize(100, 50))
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/view.py]: shall we delete this comment?

File: scripts/Muon/GUI/Common/load_file_widget/view.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.file_path_edit = QtGui.QLineEdit(self)
+
+        size_policy = QtGui.QSizePolicy(QtGui.QSizePolicy.Minimum, QtGui.QSizePolicy.Fixed)
+        size_policy.setHorizontalStretch(0)
+        size_policy.setVerticalStretch(0)
+        size_policy.setHeightForWidth(self.file_path_edit.sizePolicy().hasHeightForWidth())
+
+        self.file_path_edit.setSizePolicy(size_policy)
+        # self.file_path_edit.setMinimumSize(QtCore.QSize(200, 50))
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_file_widget/view.py]: same as above

File: scripts/test/Muon/load_file_widget/loadfile_model_test.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+if sys.version_info.major < 2:
+    from unittest import mock
+else:
+    import mock
+
+from Muon.GUI.Common.load_file_widget.model import BrowseFileWidgetModel
+from Muon.GUI.Common.muon_load_data import MuonLoadData
+
+
+class IteratorWithException:
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_file_widget/loadfile_model_test.py]: this is used in other tests. Is it possible to have it once in its own file?

File: scripts/test/Muon/load_file_widget/loadfile_model_test.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        six.assertCountEqual(self, self.model.loaded_workspaces, [[1], [2], [3]])
+        six.assertCountEqual(self, self.model.loaded_filenames, files)
+        six.assertCountEqual(self, self.model.loaded_runs, [19489, 19490, 19491])
+
+    def test_model_is_cleared_correctly(self):
+        files = [r'EMU00019489.nxs', r'EMU00019490.nxs', r'EMU00019491.nxs']
+        load_return_vals = [([1 + i], 19489 + i, filename) for i, filename in enumerate(files)]
+        self.mock_load_function(files, load_return_vals)
+
+        self.model.execute()
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_file_widget/loadfile_model_test.py]: check it loads something first

File: scripts/test/Muon/load_file_widget/loadfile_presenter_single_file_test.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+    @run_test_with_and_without_threading
+    def test_buttons_disabled_while_load_thread_running(self):
+        self.mock_browse_button_to_return_files(["file.nxs"])
+
+        self.presenter.on_browse_button_clicked()
+        self.wait_for_thread(self.presenter._load_thread)
+
+        self.assertEqual(self.view.disable_load_buttons.call_count, 1)
+        self.assertEqual(self.view.enable_load_buttons.call_count, 1)
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_file_widget/loadfile_presenter_single_file_test.py]: check it has loaded something

File: scripts/test/Muon/load_file_widget/loadfile_presenter_single_file_test.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+
+    @run_test_with_and_without_threading
+    def test_buttons_enabled_after_load_even_if_load_thread_throws(self):
+        self.model.execute = mock.Mock(side_effect=self.load_failure)
+
+        self.presenter.on_browse_button_clicked()
+        self.wait_for_thread(self.presenter._load_thread)
+
+        self.assertEqual(self.view.disable_load_buttons.call_count,
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_file_widget/loadfile_presenter_single_file_test.py]: check it has loaded something and that the disable/enable were called at least once

**Comment by AnthonyLim23** [comment]
I am happy once the builds pass

**Review by AnthonyLim23 — DISMISSED** [review]

**Comment by AndreiSavici** [comment]
Is this still in progress?

**Review by AnthonyLim23 — APPROVED** [review]
unrelated failure

**Pull request merged**